pkg/logger/logger.go

```diff
@@ -0,0 +1,146 @@
+requestID's thread is closed as a sync map is used internally.
+Child loggers can be created from a parent logger using the WithFields method, this child logger
+will append the provided fields to all logs, whilst maintaining a reference to the top level
+request fields pool.
+
+Example:
+
+WrappedLogger := NewLogger(myLogger)
+WrappedLogger.WriteFields("my-id", zap.String("foo", "bar"))
+WrappedLogger2 := WrappedLogger.WithFields(zap.String("ping", "pong"))
+
+WrappedLogger.DebugWithID("myID", "my log line")
+	=> {"level":"debug","requestID":"myID","foo":"bar","msg":"my log line""}
+
+WrappedLogger2.DebugWithID("myID", "my log line")
+	=> {"level":"debug","requestID":"myID","foo":"bar","ping":"pong","msg":"my log line""}
+
+WrappedLogger2.WriteFields("myID", zap.String("food", "bars"))
+
+WrappedLogger.DebugWithID("myID", "my log line")
+	=> {"level":"debug","requestID":"myID","foo":"bar","food":"bars","msg":"my log line""}
+*/
+
+const RequestIDFieldName = "requestID"
+
+type Logger struct {
+	requestFields *sync.Map
+	Logger        *zap.Logger
+	fields        []zap.Field
+}
+
+func (l *Logger) DebugWithID(reqID string, msg string, fields ...zap.Field) {
+	fields = append(fields, l.getFieldsForLog(reqID)...)
+	l.Logger.Debug(msg, fields...)
+}
+
+func (l *Logger) Debug(msg string, fields ...zap.Field) {
+	fields = append(fields, l.fields...)
+	l.Logger.Debug(msg, fields...)
+}
+
+func (l *Logger) InfoWithID(reqID string, msg string, fields ...zap.Field) {
+	fields = append(fields, l.getFieldsForLog(reqID)...)
+	l.Logger.Info(msg, fields...)
+}
+
+func (l *Logger) Info(msg string, fields ...zap.Field) {
+	fields = append(fields, l.fields...)
+	l.Logger.Info(msg, fields...)
+}
+
+func (l *Logger) WarnWithID(reqID string, msg string, fields ...zap.Field) {
+	fields = append(fields, l.getFieldsForLog(reqID)...)
+	l.Logger.Warn(msg, fields...)
+}
+
+func (l *Logger) Warn(msg string, fields ...zap.Field) {
+	fields = append(fields, l.fields...)
+	l.Logger.Warn(msg, fields...)
+}
+
+func (l *Logger) ErrorWithID(reqID string, msg string, fields ...zap.Field) {
+	fields = append(fields, l.getFieldsForLog(reqID)...)
+	l.Logger.Error(msg, fields...)
+}
+
+func (l *Logger) Error(msg string, fields ...zap.Field) {
+	fields = append(fields, l.fields...)
+	l.Logger.Error(msg, fields...)
+}
+
+func (l *Logger) FatalWithID(reqID string, msg string, fields ...zap.Field) {
+	fields = append(fields, l.getFieldsForLog(reqID)...)
+	l.Logger.Fatal(msg, fields...)
+}
+
+func (l *Logger) Fatal(msg string, fields ...zap.Field) {
+	fields = append(fields, l.fields...)
+	l.Logger.Debug(msg, fields...)
+}
+
+// WriteFields adds field key and value pairs to the highest level Logger, they will be applied to all
+// subsequent log calls using the matching requestID
+func (l *Logger) WriteFields(reqID string, fields ...zap.Field) {
+	res := append(l.getFields(reqID), fields...)
+	l.requestFields.Store(reqID, res)
```

Comment:
Are there any concerns with the performance of `sync.Map` with a high number of concurrent read/writes?

Reply:
If it is, I wonder whether we might benefit from https://pkg.go.dev/go.uber.org/zap#Logger.Check to avoid retrieving fields when unnecessary

Reply:
While switching to a mutex may be more efficient performance wise I believe this to be a suitable use case for the `sync.Map` as per their documentation [here](https://pkg.go.dev/sync#Map). I believe we fit use case (2). 

That being said we can alter the structure of the logger to be more inline with the following:
```
type Logger struct {
	requestFields map[string]RequestFields
	Logger        *zap.Logger
	fields        []zap.Field
	mu            *sync.Mutex
}

type RequestFields struct {
	mu     *sync.Mutex
	fields []zap.Field
}
```
Top level mu for locking the map on initial write / delete then a mutex for the reading and writing of the request specific fields? Is this what you had in mind

Reply:
I agree that sync.Map is suited to the current pattern, I was just wondering whether the performance impact is large enough to consider an alternative pattern (perhaps passing a logger instance down the request stack).
Probably a conversation for another time if performance issues arise but I thought it worth raising at least.
I do think we'd benefit from using https://pkg.go.dev/go.uber.org/zap#Logger.Check to avoid retrieving fields unless strictly necessary.

Reply:
I have implemented `Logger.Check` to wrap the field fetching - I opted to avoid passing loggers through the request stack to keep interfaces clean across the different services, however I'm happy to refactor if we feel like its required for performance